python/lsst/dax/apdb/apdbUpdateRecord.py

```diff
@@ -0,0 +1,259 @@
+            return klass(update_time_ns=update_time_ns, update_order=update_order, **kw)
+        else:
+            raise ValueError(f"Unknown update type: {update_type}")
+
+    def to_json(self) -> str:
+        data = dataclasses.asdict(self)
+        # These fields are stored separately.
+        data.pop("update_time_ns")
+        data.pop("update_order")
+        data["update_type"] = self.update_type
+        return json.dumps(data)
+
+
+@dataclass(kw_only=True)
+class ApdbReassignDiaSourceRecord(ApdbUpdateRecord, update_type="reassign_diasource"):
+    """Update record representing re-assignment of DIASource to SSObject."""
+
+    diaSourceId: int
+    """ID of DIASource record."""
+
+    diaObjectId: int
+    """ID of associated DIAObject record."""
+
+    ssObjectId: int
+    """ID of SSObject to re-associate to."""
+
+    ssObjectReassocTimeMjdTai: float
+    """Time when DIASource was re-associated from DIAObject to SSObject."""
+
+    ra: float
+    """DIASource ra, not required to be exact, but needs to be close to the
+    database record.
+    """
+
+    dec: float
+    """DIASource dec, not required to be exact, but needs to be close to the
+    database record.
+    """
+
+    apdb_table: ClassVar[ApdbTables] = ApdbTables.DiaSource
+
+
+@dataclass(kw_only=True)
+class ApdbWithdrawDiaSourceRecord(ApdbUpdateRecord, update_type="withdraw_diasource"):
+    """Update record representing withdrawal of DIASource."""
+
+    diaSourceId: int
+    """ID of DIASource record."""
+
+    diaObjectId: int | None
+    """ID of associated DIAObject record or None if it is associated to
+    SSObject.
+    """
+
+    timeWithdrawnMjdTai: float
+    """Time when this record was marked invalid."""
+
+    ra: float
+    """DIASource ra, not required to be exact, but needs to be close to the
+    database record.
+    """
+
+    dec: float
+    """DIASource dec, not required to be exact, but needs to be close to the
+    database record.
+    """
+
+    apdb_table: ClassVar[ApdbTables] = ApdbTables.DiaSource
+
+
+@dataclass(kw_only=True)
+class ApdbWithdrawDiaForcedSourceRecord(ApdbUpdateRecord, update_type="withdraw_diaforcedsource"):
+    """Update record representing withdrawal of DIAForcedSource."""
+
+    diaObjectId: int
+    """ID of DIAObject for withdrawn DIAForcedSource."""
+
+    visit: int
+    """Visit ID of DIAForcedSource."""
+
+    detector: int
+    """Detector ID of DIAForcedSource."""
+
+    timeWithdrawnMjdTai: float
+    """Time when this record was marked invalid."""
+
+    ra: float
+    """DIAForcedSource ra, not required to be exact, but needs to be close to
+    the database record.
+    """
+
+    dec: float
+    """DIAForcedSource dec, not required to be exact, but needs to be close to
+    the database record.
+    """
+
+    apdb_table: ClassVar[ApdbTables] = ApdbTables.DiaForcedSource
+
+
+@dataclass(kw_only=True)
+class ApdbCloseDiaObjectValidityRecord(ApdbUpdateRecord, update_type="close_diaobject_validity"):
```

Comment:
Will these types of updates happen one by one and independently of the replication blocks?

I thought the validity end would be assigned automatically when a `DiaObject` got a new version, which is ingested as a full row. Or at least my plan had been that this could be done in bulk during data ingestion into BigQuery when processing multiple replica chunks from staging to production.

Reply:
Covered in APDB meeting - we might want to "retire" records by closing their validity interval and not introducing a new record.